File: backend/app/schemas/configuration.py
from __future__ import annotations

from pydantic import BaseModel, ConfigDict, Field


class ResponseModel(BaseModel):
    model_config = ConfigDict(from_attributes=True)


class HomeSpaceResponse(ResponseModel):
    id: int
    name: str
    description: str
    is_active: bool


class HomeSpaceUpdate(BaseModel):
    name: str = Field(min_length=1, max_length=120)
    description: str = Field(default="", max_length=255)


class ResidenceCreate(BaseModel):
    name: str = Field(min_length=1, max_length=120)
    description: str = Field(default="", max_length=255)
    address: str = Field(default="", max_length=255)
    sort_order: int = 0


class ResidenceUpdate(BaseModel):
    name: str = Field(min_length=1, max_length=120)
    description: str = Field(default="", max_length=255)
    address: str = Field(default="", max_length=255)
    sort_order: int = 0
    is_active: bool = True


class ResidenceResponse(ResidenceUpdate):
    model_config = ConfigDict(from_attributes=True)

    id: int


class LocationNodeCreate(BaseModel):
    residence_id: int
    parent_id: int | None = None
    name: str = Field(min_length=1, max_length=120)
    node_type: str = Field(default="area", max_length=40)
    icon: str = Field(default="", max_length=60)
    sort_order: int = 0
    note: str = Field(default="", max_length=255)


class LocationNodeUpdate(BaseModel):
    parent_id: int | None = None
    name: str = Field(min_length=1, max_length=120)
    node_type: str = Field(default="area", max_length=40)
    icon: str = Field(default="", max_length=60)
    sort_order: int = 0
    note: str = Field(default="", max_length=255)
    is_active: bool = True


class LocationNodeResponse(LocationNodeUpdate):
    model_config = ConfigDict(from_attributes=True)

    id: int
    residence_id: int
    children: list[LocationNodeResponse] = Field(default_factory=list)


class FamilyMemberCreate(BaseModel):
    name: str = Field(min_length=1, max_length=120)
    relation: str = Field(default="", max_length=80)
    phone: str = Field(default="", max_length=80)
    note: str = Field(default="", max_length=255)


class FamilyMemberUpdate(FamilyMemberCreate):
    is_active: bool = True


class FamilyMemberResponse(FamilyMemberUpdate):
    model_config = ConfigDict(from_attributes=True)

    id: int
    user_id: int | None


class CategoryCreate(BaseModel):
    parent_id: int | None = None
    code: str = Field(min_length=1, max_length=80)
    name: str = Field(min_length=1, max_length=120)
    icon: str = Field(default="", max_length=60)
    sort_order: int = 0


class CategoryUpdate(BaseModel):
    parent_id: int | None = None
    name: str = Field(min_length=1, max_length=120)
    icon: str = Field(default="", max_length=60)
    sort_order: int = 0
    is_active: bool = True


class CategoryResponse(CategoryUpdate):
    model_config = ConfigDict(from_attributes=True)

    id: int
    code: str
    attribute_definitions: list[AttributeDefinitionResponse] = Field(default_factory=list)
    children: list[CategoryResponse] = Field(default_factory=list)


class AttributeDefinitionCreate(BaseModel):
    category_id: int
    key: str = Field(min_length=1, max_length=80)
    name: str = Field(min_length=1, max_length=120)
    field_type: str = Field(min_length=1, max_length=40)
    default_value: str = ""
    privacy_level: str = "normal"
    is_required: bool = False
    is_filterable: bool = False
    sort_order: int = 0


class AttributeDefinitionUpdate(BaseModel):
    name: str = Field(min_length=1, max_length=120)
    field_type: str = Field(min_length=1, max_length=40)
    default_value: str = ""
    privacy_level: str = "normal"
    is_required: bool = False
    is_filterable: bool = False
    sort_order: int = 0
    is_active: bool = True


class AttributeOptionCreate(BaseModel):
    definition_id: int
    label: str = Field(min_length=1, max_length=120)
    value: str = Field(min_length=1, max_length=120)
    sort_order: int = 0


class AttributeOptionUpdate(BaseModel):
    label: str = Field(min_length=1, max_length=120)
    sort_order: int = 0
    is_active: bool = True


class AttributeOptionResponse(AttributeOptionUpdate):
    model_config = ConfigDict(from_attributes=True)

    id: int
    definition_id: int
    value: str


class AttributeDefinitionResponse(AttributeDefinitionUpdate):
    model_config = ConfigDict(from_attributes=True)

    id: int
    category_id: int
    key: str
    options: list[AttributeOptionResponse] = Field(default_factory=list)


class ItemStatusCreate(BaseModel):
    code: str = Field(min_length=1, max_length=80)
    name: str = Field(min_length=1, max_length=120)
    semantic: str = Field(min_length=1, max_length=80)
    sort_order: int = 0


class ItemStatusUpdate(BaseModel):
    name: str = Field(min_length=1, max_length=120)
    semantic: str = Field(min_length=1, max_length=80)
    sort_order: int = 0
    is_active: bool = True


class ItemStatusResponse(ItemStatusUpdate):
    model_config = ConfigDict(from_attributes=True)

    id: int
    code: str
    is_system: bool


class DictionaryGroupCreate(BaseModel):
    code: str = Field(min_length=1, max_length=80)
    name: str = Field(min_length=1, max_length=120)


class DictionaryOptionResponse(ResponseModel):
    id: int
    group_id: int
    label: str
    value: str
    sort_order: int
    is_active: bool


class DictionaryGroupResponse(ResponseModel):
    id: int
    code: str
    name: str
    is_system: bool
    is_active: bool
    options: list[DictionaryOptionResponse] = Field(default_factory=list)


class DictionaryOptionCreate(BaseModel):
    group_id: int
    label: str = Field(min_length=1, max_length=120)
    value: str = Field(min_length=1, max_length=120)
    sort_order: int = 0


class ConfigBootstrapResponse(BaseModel):
    home_space: HomeSpaceResponse
    residences: list[ResidenceResponse]
    location_tree: list[LocationNodeResponse]
    family_members: list[FamilyMemberResponse]
    categories: list[CategoryResponse]
    item_statuses: list[ItemStatusResponse]
    dictionary_groups: list[DictionaryGroupResponse]

File: frontend/src/api/configuration.ts
import { apiClient } from './client'

export type PrivacyLevel = 'normal' | 'sensitive'
export type AttributeFieldType =
  | 'text'
  | 'long_text'
  | 'number'
  | 'money'
  | 'date'
  | 'datetime'
  | 'single_select'
  | 'multi_select'
  | 'boolean'
  | 'url'
  | 'attachment'
  | 'reminder_date'

export interface HomeSpace {
  id: number
  name: string
  description: string
  is_active: boolean
}

export interface Residence {
  id: number
  name: string
  description: string
  address: string
  sort_order: number
  is_active: boolean
}

export interface LocationNode {
  id: number
  residence_id: number
  parent_id: number | null
  name: string
  node_type: string
  icon: string
  sort_order: number
  note: string
  is_active: boolean
  children: LocationNode[]
}

export interface FamilyMember {
  id: number
  name: string
  relation: string
  phone: string
  note: string
  is_active: boolean
  user_id: number | null
}

export interface Category {
  id: number
  parent_id: number | null
  code: string
  name: string
  icon: string
  sort_order: number
  is_active: boolean
  attribute_definitions: AttributeDefinition[]
  children: Category[]
}

export interface AttributeOption {
  id: number
  definition_id: number
  label: string
  value: string
  sort_order: number
  is_active: boolean
}

export interface AttributeDefinition {
  id: number
  category_id: number
  key: string
  name: string
  field_type: AttributeFieldType
  default_value: string
  privacy_level: PrivacyLevel
  is_required: boolean
  is_filterable: boolean
  sort_order: number
  is_active: boolean
  options: AttributeOption[]
}

export interface ItemStatus {
  id: number
  code: string
  name: string
  semantic: string
  sort_order: number
  is_active: boolean
  is_system: boolean
}

export interface DictionaryOption {
  id: number
  group_id: number
  label: string
  value: string
  sort_order: number
  is_active: boolean
}

export interface DictionaryGroup {
  id: number
  code: string
  name: string
  is_system: boolean
  is_active: boolean
  options: DictionaryOption[]
}

export interface ConfigBootstrap {
  home_space: HomeSpace
  residences: Residence[]
  location_tree: LocationNode[]
  family_members: FamilyMember[]
  categories: Category[]
  item_statuses: ItemStatus[]
  dictionary_groups: DictionaryGroup[]
}

export interface ResidenceCreate {
  name: string
  description?: string
  address?: string
  sort_order?: number
}

export interface ResidenceUpdate {
  name: string
  description: string
  address: string
  sort_order: number
  is_active: boolean
}

export interface LocationNodeCreate {
  residence_id: number
  parent_id?: number | null
  name: string
  node_type?: string
  icon?: string
  sort_order?: number
  note?: string
}

export interface LocationNodeUpdate {
  parent_id: number | null
  name: string
  node_type: string
  icon: string
  sort_order: number
  note: string
  is_active: boolean
}

export interface FamilyMemberCreate {
  name: string
  relation?: string
  phone?: string
  note?: string
}

export interface FamilyMemberUpdate {
  name: string
  relation: string
  phone: string
  note: string
  is_active: boolean
}

export interface CategoryCreate {
  parent_id?: number | null
  code: string
  name: string
  icon?: string
  sort_order?: number
}

export interface CategoryUpdate {
  parent_id: number | null
  name: string
  icon: string
  sort_order: number
  is_active: boolean
}

export interface AttributeDefinitionCreate {
  category_id: number
  key: string
  name: string
  field_type: AttributeFieldType
  default_value?: string
  privacy_level?: PrivacyLevel
  is_required?: boolean
  is_filterable?: boolean
  sort_order?: number
}

export interface AttributeDefinitionUpdate {
  name: string
  field_type: AttributeFieldType
  default_value: string
  privacy_level: PrivacyLevel
  is_required: boolean
  is_filterable: boolean
  sort_order: number
  is_active: boolean
}

export interface AttributeOptionCreate {
  definition_id: number
  label: string
  value: string
  sort_order?: number
}

export interface AttributeOptionUpdate {
  label: string
  sort_order: number
  is_active: boolean
}

export interface ItemStatusCreate {
  code: string
  name: string
  semantic: string
  sort_order?: number
}

export interface ItemStatusUpdate {
  name: string
  semantic: string
  sort_order: number
  is_active: boolean
}

export async function fetchConfigBootstrapApi(): Promise<ConfigBootstrap> {
  const response = await apiClient.get<ConfigBootstrap>('/config/bootstrap')
  return response.data
}

export async function createResidenceApi(payload: ResidenceCreate): Promise<Residence> {
  const response = await apiClient.post<Residence>('/config/residences', payload)
  return response.data
}

export async function updateResidenceApi(
  residenceId: number,
  payload: ResidenceUpdate
): Promise<Residence> {
  const response = await apiClient.patch<Residence>(`/config/residences/${residenceId}`, payload)
  return response.data
}

export async function createLocationNodeApi(payload: LocationNodeCreate): Promise<LocationNode> {
  const response = await apiClient.post<LocationNode>('/config/location-nodes', payload)
  return response.data
}

export async function updateLocationNodeApi(
  nodeId: number,
  payload: LocationNodeUpdate
): Promise<LocationNode> {
  const response = await apiClient.patch<LocationNode>(`/config/location-nodes/${nodeId}`, payload)
  return response.data
}

export async function createFamilyMemberApi(payload: FamilyMemberCreate): Promise<FamilyMember> {
  const response = await apiClient.post<FamilyMember>('/config/family-members', payload)
  return response.data
}

export async function updateFamilyMemberApi(
  memberId: number,
  payload: FamilyMemberUpdate
): Promise<FamilyMember> {
  const response = await apiClient.patch<FamilyMember>(`/config/family-members/${memberId}`, payload)
  return response.data
}

export async function createCategoryApi(payload: CategoryCreate): Promise<Category> {
  const response = await apiClient.post<Category>('/config/categories', payload)
  return response.data
}

export async function updateCategoryApi(
  categoryId: number,
  payload: CategoryUpdate
): Promise<Category> {
  const response = await apiClient.patch<Category>(`/config/categories/${categoryId}`, payload)
  return response.data
}

export async function createAttributeDefinitionApi(
  payload: AttributeDefinitionCreate
): Promise<AttributeDefinition> {
  const response = await apiClient.post<AttributeDefinition>('/config/attribute-definitions', payload)
  return response.data
}

export async function updateAttributeDefinitionApi(
  definitionId: number,
  payload: AttributeDefinitionUpdate
): Promise<AttributeDefinition> {
  const response = await apiClient.patch<AttributeDefinition>(
    `/config/attribute-definitions/${definitionId}`,
    payload
  )
  return response.data
}

export async function createAttributeOptionApi(
  payload: AttributeOptionCreate
): Promise<AttributeOption> {
  const response = await apiClient.post<AttributeOption>('/config/attribute-options', payload)
  return response.data
}

export async function updateAttributeOptionApi(
  optionId: number,
  payload: AttributeOptionUpdate
): Promise<AttributeOption> {
  const response = await apiClient.patch<AttributeOption>(`/config/attribute-options/${optionId}`, payload)
  return response.data
}

export async function createItemStatusApi(payload: ItemStatusCreate): Promise<ItemStatus> {
  const response = await apiClient.post<ItemStatus>('/config/item-statuses', payload)
  return response.data
}

export async function updateItemStatusApi(
  statusId: number,
  payload: ItemStatusUpdate
): Promise<ItemStatus> {
  const response = await apiClient.patch<ItemStatus>(`/config/item-statuses/${statusId}`, payload)
  return response.data
}
